File: web-app/src/services/aiService.js
import api from './api';

const AI_BASE_PATH = '/ai';

export const aiService = {
  prioritize: async (tasks) => {
    const response = await api.post(`${AI_BASE_PATH}/prioritize`, { tasks });
    return response.data;
  },

  assistWrite: async ({ title, category, description }) => {
    const response = await api.post(`${AI_BASE_PATH}/assist-write`, {
      title,
      category,
      description,
    });
    return response.data;
  },

  predictTime: async (task) => {
    const response = await api.post(`${AI_BASE_PATH}/predict-time`, {
      task,
    });
    return response.data;
  },

  planDay: async (tasks, workStart, workEnd) => {
    const response = await api.post(`${AI_BASE_PATH}/plan-day`, {
      tasks,
      work_start: workStart,
      work_end: workEnd,
    });
    return response.data;
  },

  detectRisks: async (tasks) => {
    const response = await api.post(`${AI_BASE_PATH}/detect-risks`, { tasks });
    return response.data;
  },

  generateReport: async (stats) => {
    const payload = {
      period: stats.period,
      total_tasks: stats.total_tasks ?? stats.totalTasks,
      completed: stats.completed,
      overdue: stats.overdue,
      average_completion_minutes:
        stats.average_completion_minutes ?? stats.averageCompletionMinutes,
      tasks_by_day: stats.tasks_by_day ?? stats.tasksByDay,
      tasks_by_period: stats.tasks_by_period ?? stats.tasksByPeriod,
    };

    const response = await api.post(`${AI_BASE_PATH}/reports`, payload);
    return response.data;
  },

  projectBreakdown: async (
    name,
    description,
    projectType,
    team,
    scope,
    taskCount,
    projectDeadline
  ) => {
    const payload = {
      name,
      description,
      project_type: projectType,
      team,
      scope,
      task_count: taskCount,
    };

    if (projectDeadline) {
      payload.project_deadline = projectDeadline;
    }

    const response = await api.post(`${AI_BASE_PATH}/project-breakdown`, payload);
    return response.data;
  },

  generateSubtasks: async (
    projectContext,
    taskTitle,
    taskDescription,
    category,
    phase,
    estimatedMinutes
  ) => {
    const response = await api.post(`${AI_BASE_PATH}/generate-subtasks`, {
      project_context: projectContext,
      task_title: taskTitle,
      task_description: taskDescription,
      category,
      phase,
      estimated_minutes: estimatedMinutes,
    });
    return response.data;
  },

  generateDependencies: async (tasks) => {
    const response = await api.post(`${AI_BASE_PATH}/generate-dependencies`, {
      tasks,
    });
    return response.data;
  },
};

export default aiService;

File: mobile/lib/services/ai_service.dart
import 'package:dio/dio.dart';

import 'api_service.dart';

class AiService {
  AiService({Dio? dio}) : _dio = dio ?? ApiService.instance.dio;

  final Dio _dio;

  Map<String, dynamic> _asMap(dynamic data) {
    if (data is Map<String, dynamic>) {
      return data;
    }

    if (data is Map) {
      return Map<String, dynamic>.from(data);
    }

    if (data is List && data.isNotEmpty && data.first is Map) {
      return Map<String, dynamic>.from(data.first as Map);
    }

    return <String, dynamic>{};
  }

  Future<Map<String, dynamic>> prioritize(List<Map<String, dynamic>> tasks) async {
    final res = await _dio.post('/ai/prioritize', data: <String, dynamic>{'tasks': tasks});
    return _asMap(res.data);
  }

  Future<Map<String, dynamic>> assistWrite({String? title, String? category, String? description}) async {
    final res = await _dio.post('/ai/assist-write', data: <String, dynamic>{
      'title': title,
      'category': category,
      'description': description,
    });
    return _asMap(res.data);
  }

  Future<Map<String, dynamic>> predictTime(Map<String, dynamic> task) async {
    final res = await _dio.post('/ai/predict-time', data: <String, dynamic>{'task': task});
    return _asMap(res.data);
  }

  Future<Map<String, dynamic>> planDay(
    List<Map<String, dynamic>> tasks,
    String workStart,
    String workEnd,
  ) async {
    final res = await _dio.post('/ai/plan-day', data: <String, dynamic>{
      'tasks': tasks,
      'work_start': workStart,
      'work_end': workEnd,
    });
    return _asMap(res.data);
  }

  Future<Map<String, dynamic>> detectRisks(List<Map<String, dynamic>> tasks) async {
    final res = await _dio.post('/ai/detect-risks', data: <String, dynamic>{'tasks': tasks});
    return _asMap(res.data);
  }

  Future<Map<String, dynamic>> generateReport(Map<String, dynamic> stats) async {
    final res = await _dio.post('/ai/reports', data: stats);
    return _asMap(res.data);
  }

  Future<Map<String, dynamic>> projectBreakdown(
    String name,
    String description,
    String projectType,
    String team,
    String scope,
    int taskCount,
    String? deadline,
  ) async {
    final payload = <String, dynamic>{
      'name': name,
      'description': description,
      'project_type': projectType,
      'team': team,
      'scope': scope,
      'task_count': taskCount,
    };
    if (deadline != null && deadline.isNotEmpty) payload['project_deadline'] = deadline;
    final res = await _dio.post('/ai/project-breakdown', data: payload);
    return _asMap(res.data);
  }

  Future<Map<String, dynamic>> generateSubtasks(
    String projectContext,
    String taskTitle,
    String taskDescription,
    String category,
    String phase,
    int estimatedMinutes,
  ) async {
    final res = await _dio.post('/ai/generate-subtasks', data: <String, dynamic>{
      'project_context': projectContext,
      'task_title': taskTitle,
      'task_description': taskDescription,
      'category': category,
      'phase': phase,
      'estimated_minutes': estimatedMinutes,
    });
    return _asMap(res.data);
  }

  Future<Map<String, dynamic>> enhanceProject(String name, String description) async {
    final res = await _dio.post('/ai/enhance-project', data: <String, dynamic>{
      'name': name,
      'description': description,
    });
    return _asMap(res.data);
  }
}
